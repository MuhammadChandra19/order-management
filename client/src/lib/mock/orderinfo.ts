export const MOCK_ORDER_INFO = [
  {
      "id": "1",
      "order_name": "PO #001-I",
      "company_name": "Roga & Kopyta",
      "customer_name": "Ivan Ivanovich",
      "product_name": "Corrugated Box",
      "order_date": "2020-01-02T15:34:12Z",
      "delivered": "5",
      "total_amount": 13.454
  },
  {
      "id": "6",
      "order_name": "PO #001-P",
      "company_name": "Pupkin & Co",
      "customer_name": "Petr Petrovich",
      "product_name": "Corrugated Box",
      "order_date": "2020-01-02T15:34:12Z",
      "delivered": "15",
      "total_amount": 23.181
  },
  {
      "id": "3",
      "order_name": "PO #003-I",
      "company_name": "Roga & Kopyta",
      "customer_name": "Ivan Ivanovich",
      "product_name": "Corrugated Box",
      "order_date": "2020-01-05T05:34:12Z",
      "delivered": "12",
      "total_amount": 1476.414
  },
  {
      "id": "8",
      "order_name": "PO #003-P",
      "company_name": "Pupkin & Co",
      "customer_name": "Petr Petrovich",
      "product_name": "Corrugated Box",
      "order_date": "2020-01-05T05:34:12Z",
      "delivered": "17",
      "total_amount": 6784.7595
  },
  {
      "id": "7",
      "order_name": "PO #002-P",
      "company_name": "Pupkin & Co",
      "customer_name": "Petr Petrovich",
      "product_name": "Corrugated Box",
      "order_date": "2020-01-15T17:34:12Z",
      "delivered": "8",
      "total_amount": 402.24
  }
]