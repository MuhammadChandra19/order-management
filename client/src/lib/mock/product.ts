export const PRODUCT_STATS = [
  {
      "id": "14",
      "order_name": "PO #009-M",
      "company_name": "ABC Corporation",
      "customer_name": "Maria Petrovna",
      "product_name": "Shipping Box",
      "order_date": "2023-04-02T18:20:12Z",
      "delivered": "7",
      "total_amount": 61.25
  },
  {
      "id": "19",
      "order_name": "PO #009-J",
      "company_name": "XYZ Ltd",
      "customer_name": "John Doe",
      "product_name": "Bubble Wrap",
      "order_date": "2023-04-02T18:20:12Z",
      "delivered": "30",
      "total_amount": 479.7
  },
  {
      "id": "17",
      "order_name": "PO #007-J",
      "company_name": "XYZ Ltd",
      "customer_name": "John Doe",
      "product_name": "Label Sticker",
      "order_date": "2023-03-15T12:45:12Z",
      "delivered": "22",
      "total_amount": 120.78
  },
  {
      "id": "11",
      "order_name": "PO #007-M",
      "company_name": "ABC Corporation",
      "customer_name": "Maria Petrovna",
      "product_name": "Box Tape",
      "order_date": "2023-03-15T12:45:12Z",
      "delivered": "10",
      "total_amount": 155
  },
  {
      "id": "13",
      "order_name": "PO #008-M",
      "company_name": "ABC Corporation",
      "customer_name": "Maria Petrovna",
      "product_name": "Label Sticker",
      "order_date": "2023-03-05T08:30:12Z",
      "delivered": "15",
      "total_amount": 825
  }
]