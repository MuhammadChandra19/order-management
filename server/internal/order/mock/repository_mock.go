// Code generated by MockGen. DO NOT EDIT.
// Source: repository.go

// Package mock is a generated GoMock package.
package mock

import (
	reflect "reflect"
	time "time"

	order "github.com/MuhammadChandra19/order-management/internal/order"
	gomock "github.com/golang/mock/gomock"
)

// MockOrderRepositoryInterface is a mock of OrderRepositoryInterface interface.
type MockOrderRepositoryInterface struct {
	ctrl     *gomock.Controller
	recorder *MockOrderRepositoryInterfaceMockRecorder
}

// MockOrderRepositoryInterfaceMockRecorder is the mock recorder for MockOrderRepositoryInterface.
type MockOrderRepositoryInterfaceMockRecorder struct {
	mock *MockOrderRepositoryInterface
}

// NewMockOrderRepositoryInterface creates a new mock instance.
func NewMockOrderRepositoryInterface(ctrl *gomock.Controller) *MockOrderRepositoryInterface {
	mock := &MockOrderRepositoryInterface{ctrl: ctrl}
	mock.recorder = &MockOrderRepositoryInterfaceMockRecorder{mock}
	return mock
}

// EXPECT returns an object that allows the caller to indicate expected use.
func (m *MockOrderRepositoryInterface) EXPECT() *MockOrderRepositoryInterfaceMockRecorder {
	return m.recorder
}

// GetOrderList mocks base method.
func (m *MockOrderRepositoryInterface) GetOrderList(search string, startDate, endDate time.Time, sortDirection string, limit, offset int) ([]*order.OrderInfo, error) {
	m.ctrl.T.Helper()
	ret := m.ctrl.Call(m, "GetOrderList", search, startDate, endDate, sortDirection, limit, offset)
	ret0, _ := ret[0].([]*order.OrderInfo)
	ret1, _ := ret[1].(error)
	return ret0, ret1
}

// GetOrderList indicates an expected call of GetOrderList.
func (mr *MockOrderRepositoryInterfaceMockRecorder) GetOrderList(search, startDate, endDate, sortDirection, limit, offset interface{}) *gomock.Call {
	mr.mock.ctrl.T.Helper()
	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "GetOrderList", reflect.TypeOf((*MockOrderRepositoryInterface)(nil).GetOrderList), search, startDate, endDate, sortDirection, limit, offset)
}
