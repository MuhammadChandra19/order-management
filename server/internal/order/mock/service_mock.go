// Code generated by MockGen. DO NOT EDIT.
// Source: service.go

// Package mock is a generated GoMock package.
package mock

import (
	reflect "reflect"

	gin "github.com/gin-gonic/gin"
	gomock "github.com/golang/mock/gomock"
)

// MockOrderServiceInterface is a mock of OrderServiceInterface interface.
type MockOrderServiceInterface struct {
	ctrl     *gomock.Controller
	recorder *MockOrderServiceInterfaceMockRecorder
}

// MockOrderServiceInterfaceMockRecorder is the mock recorder for MockOrderServiceInterface.
type MockOrderServiceInterfaceMockRecorder struct {
	mock *MockOrderServiceInterface
}

// NewMockOrderServiceInterface creates a new mock instance.
func NewMockOrderServiceInterface(ctrl *gomock.Controller) *MockOrderServiceInterface {
	mock := &MockOrderServiceInterface{ctrl: ctrl}
	mock.recorder = &MockOrderServiceInterfaceMockRecorder{mock}
	return mock
}

// EXPECT returns an object that allows the caller to indicate expected use.
func (m *MockOrderServiceInterface) EXPECT() *MockOrderServiceInterfaceMockRecorder {
	return m.recorder
}

// GetOrderList mocks base method.
func (m *MockOrderServiceInterface) GetOrderList(c *gin.Context) {
	m.ctrl.T.Helper()
	m.ctrl.Call(m, "GetOrderList", c)
}

// GetOrderList indicates an expected call of GetOrderList.
func (mr *MockOrderServiceInterfaceMockRecorder) GetOrderList(c interface{}) *gomock.Call {
	mr.mock.ctrl.T.Helper()
	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "GetOrderList", reflect.TypeOf((*MockOrderServiceInterface)(nil).GetOrderList), c)
}
