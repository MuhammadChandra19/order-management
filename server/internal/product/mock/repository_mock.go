// Code generated by MockGen. DO NOT EDIT.
// Source: repository.go

// Package mock is a generated GoMock package.
package mock

import (
	reflect "reflect"

	product "github.com/MuhammadChandra19/order-management/internal/product"
	gomock "github.com/golang/mock/gomock"
)

// MockProductRepositoryInterface is a mock of ProductRepositoryInterface interface.
type MockProductRepositoryInterface struct {
	ctrl     *gomock.Controller
	recorder *MockProductRepositoryInterfaceMockRecorder
}

// MockProductRepositoryInterfaceMockRecorder is the mock recorder for MockProductRepositoryInterface.
type MockProductRepositoryInterfaceMockRecorder struct {
	mock *MockProductRepositoryInterface
}

// NewMockProductRepositoryInterface creates a new mock instance.
func NewMockProductRepositoryInterface(ctrl *gomock.Controller) *MockProductRepositoryInterface {
	mock := &MockProductRepositoryInterface{ctrl: ctrl}
	mock.recorder = &MockProductRepositoryInterfaceMockRecorder{mock}
	return mock
}

// EXPECT returns an object that allows the caller to indicate expected use.
func (m *MockProductRepositoryInterface) EXPECT() *MockProductRepositoryInterfaceMockRecorder {
	return m.recorder
}

// GetProductSalesStats mocks base method.
func (m *MockProductRepositoryInterface) GetProductSalesStats() ([]product.ProductSalesStats, error) {
	m.ctrl.T.Helper()
	ret := m.ctrl.Call(m, "GetProductSalesStats")
	ret0, _ := ret[0].([]product.ProductSalesStats)
	ret1, _ := ret[1].(error)
	return ret0, ret1
}

// GetProductSalesStats indicates an expected call of GetProductSalesStats.
func (mr *MockProductRepositoryInterfaceMockRecorder) GetProductSalesStats() *gomock.Call {
	mr.mock.ctrl.T.Helper()
	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "GetProductSalesStats", reflect.TypeOf((*MockProductRepositoryInterface)(nil).GetProductSalesStats))
}
