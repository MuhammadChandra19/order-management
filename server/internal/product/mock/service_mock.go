// Code generated by MockGen. DO NOT EDIT.
// Source: service.go

// Package mock is a generated GoMock package.
package mock

import (
	reflect "reflect"

	gin "github.com/gin-gonic/gin"
	gomock "github.com/golang/mock/gomock"
)

// MockProductServiceInterface is a mock of ProductServiceInterface interface.
type MockProductServiceInterface struct {
	ctrl     *gomock.Controller
	recorder *MockProductServiceInterfaceMockRecorder
}

// MockProductServiceInterfaceMockRecorder is the mock recorder for MockProductServiceInterface.
type MockProductServiceInterfaceMockRecorder struct {
	mock *MockProductServiceInterface
}

// NewMockProductServiceInterface creates a new mock instance.
func NewMockProductServiceInterface(ctrl *gomock.Controller) *MockProductServiceInterface {
	mock := &MockProductServiceInterface{ctrl: ctrl}
	mock.recorder = &MockProductServiceInterfaceMockRecorder{mock}
	return mock
}

// EXPECT returns an object that allows the caller to indicate expected use.
func (m *MockProductServiceInterface) EXPECT() *MockProductServiceInterfaceMockRecorder {
	return m.recorder
}

// GetProductSalesStats mocks base method.
func (m *MockProductServiceInterface) GetProductSalesStats(c *gin.Context) {
	m.ctrl.T.Helper()
	m.ctrl.Call(m, "GetProductSalesStats", c)
}

// GetProductSalesStats indicates an expected call of GetProductSalesStats.
func (mr *MockProductServiceInterfaceMockRecorder) GetProductSalesStats(c interface{}) *gomock.Call {
	mr.mock.ctrl.T.Helper()
	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "GetProductSalesStats", reflect.TypeOf((*MockProductServiceInterface)(nil).GetProductSalesStats), c)
}
