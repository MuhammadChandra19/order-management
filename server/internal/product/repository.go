package product

import "database/sql"

//go:generate mockgen -source repository.go -destination mock/repository_mock.go -package=mock

type ProductSalesStats struct {
	ProductName       string  `json:"product_name"`
	TotalQuantitySold int     `json:"total_quantity_sold"`
	TotalAmount       float64 `json:"total_amount"`
}

type ProductRepositoryInterface interface {
	GetProductSalesStats() ([]*ProductSalesStats, error)
}

type repository struct {
	db *sql.DB
}

// retrieve the 5 most sold products along with their corresponding total quantities sold and total amounts
func (r *repository) GetProductSalesStats() ([]*ProductSalesStats, error) {
	query := `
		SELECT 
			oi.product AS product_name,
			SUM(d.delivered_quantity) AS total_quantity_sold,
			SUM(oi.price_per_unit * d.delivered_quantity) AS total_amount
		FROM 
			order_items oi
		JOIN 
			deliveries d ON oi.id = d.order_item_id
		GROUP BY 
			oi.product
		ORDER BY 
			total_quantity_sold DESC, 
			total_amount DESC
		LIMIT 5
	`

	rows, err := r.db.Query(query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var productsSalesStats []*ProductSalesStats
	for rows.Next() {
		var productSalesStats ProductSalesStats
		err := rows.Scan(
			&productSalesStats.ProductName,
			&productSalesStats.TotalQuantitySold,
			&productSalesStats.TotalAmount,
		)
		if err != nil {
			return nil, err
		}
		productsSalesStats = append(productsSalesStats, &productSalesStats)
	}

	return productsSalesStats, nil
}

func NewProductRepository(db *sql.DB) ProductRepositoryInterface {
	return &repository{
		db: db,
	}
}
