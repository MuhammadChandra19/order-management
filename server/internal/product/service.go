package product

//go:generate mockgen -source service.go -destination mock/service_mock.go -package=mock
import (
	"net/http"

	"github.com/gin-gonic/gin"
)

// ProductServiceInterface defines the interface for product-related services.
type ProductServiceInterface interface {
	// GetProductSalesStats retrieves product sales statistics.
	//
	// Parameters:
	//
	// - c (*gin.Context): Context object for Gin HTTP request.
	//
	// Returns:
	// None.
	GetProductSalesStats(c *gin.Context)
}

type service struct {
	productRepo ProductRepositoryInterface
}

func (s *service) GetProductSalesStats(c *gin.Context) {
	products, err := s.productRepo.GetProductSalesStats()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, products)
}

// NewProductService creates a new instance of ProductServiceInterface.
//
// Parameters:
//
// - productRepo (ProductRepositoryInterface): Product repository interface.
//
// Returns:
//
// - (ProductServiceInterface): New instance of ProductServiceInterface.
func NewProductService(productRepo ProductRepositoryInterface) ProductServiceInterface {
	return &service{
		productRepo: productRepo,
	}
}
