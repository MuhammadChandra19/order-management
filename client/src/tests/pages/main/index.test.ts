import { PRODUCT_STATS } from '@/lib/mock/product';
import { render, waitFor } from '@testing-library/vue';
import { Mock } from 'vitest';
import MainPage from '@/pages/Main/index.vue'

global.fetch = vi.fn(() =>
  Promise.resolve({
    json: () => 
      Promise.resolve(PRODUCT_STATS)
  })
) as Mock

const renderComponent = () =>  render(MainPage)

describe('Main page', () => {
  afterEach(() => {
    vi.restoreAllMocks()
  })
  test('Should render page correctly',async () => {
    const { container, findAllByTestId } = renderComponent()
    await waitFor(() => {
      expect(global.fetch)
        .toHaveBeenLastCalledWith("http://localhost:8080/api/product-sale-stats")
    })

    const productRows = await findAllByTestId('product-sales-rows')
    expect(productRows).toBeDefined()
    expect(container).toMatchSnapshot()
  })
})
