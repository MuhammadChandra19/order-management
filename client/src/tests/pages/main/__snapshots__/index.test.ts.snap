// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Main page > Should render page correctly 1`] = `
<div>
  <div
    class="m-auto flex justify-between"
  >
    <div
      class="rounded-lg border bg-card text-card-foreground shadow-sm max-w-xl p-2"
    >
      
      <div
        class="flex flex-col space-y-1.5 p-6"
      >
        
        <h3
          class="text-2xl font-semibold leading-none tracking-tighter"
        >
          
          Product sale stats
          
        </h3>
        <p
          class="text-sm text-muted-foreground"
        >
          
          Show Top 5 Product sales
          
        </p>
        
      </div>
      <div
        class="p-6 pt-0"
      >
        
        <div
          class="w-full overflow-auto"
        >
          <table
            class="w-full caption-bottom text-sm"
          >
            
            <thead
              class="[&_tr]:border-b"
            >
              
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
              >
                
                <th
                  class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                >
                  
                  Product Name
                  
                </th>
                <th
                  class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                >
                  
                  Total Quantity Sold
                  
                </th>
                <th
                  class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                >
                  
                  Total Amount
                  
                </th>
                
              </tr>
              
            </thead>
            <tbody
              class="[&_tr:last-child]:border-0"
            >
              
              
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
                data-testid="product-sales-rows"
              >
                
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
                >
                  
                  Shipping Box
                  
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  
                  
                  
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
                >
                  
                  $61.25
                  
                </td>
                
              </tr>
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
                data-testid="product-sales-rows"
              >
                
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
                >
                  
                  Bubble Wrap
                  
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  
                  
                  
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
                >
                  
                  $479.7
                  
                </td>
                
              </tr>
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
                data-testid="product-sales-rows"
              >
                
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
                >
                  
                  Label Sticker
                  
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  
                  
                  
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
                >
                  
                  $120.78
                  
                </td>
                
              </tr>
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
                data-testid="product-sales-rows"
              >
                
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
                >
                  
                  Box Tape
                  
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  
                  
                  
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
                >
                  
                  $155
                  
                </td>
                
              </tr>
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
                data-testid="product-sales-rows"
              >
                
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
                >
                  
                  Label Sticker
                  
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  
                  
                  
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
                >
                  
                  $825
                  
                </td>
                
              </tr>
              
              
            </tbody>
            
          </table>
        </div>
        
      </div>
      
    </div>
  </div>
</div>
`;
